### PR TITLE
Add Identity Infrastructure

### LINKED ISSUE #35: Add Authentication and Account Management infrastructure
### Description
Create `AuthService`

### Requirements
- project should be runnable
- should allow registering and logging in to system without exceptions.
- test all the dependent services

### Deliverables
- `SignUpDetails`, `SignInDetails` entities
- mapping between `SignUpDetails` and `Client` models
- register all necessary infrastructure to DI
- `AuthController` controller with the following actions
    - `SignUp` - to sign up a new user
    - `SignIn` - to sign in a new user

### PR DESCRIPTION
**Pull request by bunyodkamalov**
closes #35

### CODE REVIEW

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Clients/Mappers/AccessTokenMapper.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using AutoMapper;
+using Feedback.Analyzer.Application.Common.Identity.Models;
+using Feedback.Analyzer.Domain.Entities;
+
+namespace Feedback.Analyzer.Api.Mappers;
```
Comment (isxoqovxasanboy) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Clients/Mappers/AccessTokenMapper.cs:5]: ## Change namespace:

'False' namespace: Feedback.Analyzer.Api.Mappers;

`True` namespace: Feedback.Analyzer.Application.Clients.Mappers;

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Identity/Services/IdentitySecurityTokenService.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+/// </summary>
+public class IdentitySecurityTokenService(
+    IAccessTokenRepository accessTokenRepository, 
+    IRefreshTokenRepository refreshTokenRepository,
+    IValidator<RefreshToken> refreshTokenValidator) : IIdentitySecurityTokenService
+{
+    public ValueTask<AccessToken> CreateAccessTokenAsync(AccessToken accessToken, bool saveChanges = true, CancellationToken cancellationToken = default)
+    {
+        return accessTokenRepository.CreateAsync(accessToken, saveChanges, cancellationToken);
+    }
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Identity/Services/IdentitySecurityTokenService.cs:19]: Add validation to access token creation method

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Settings/RequestUserContextSettings.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+namespace Feedback.Analyzer.Application.Common.Settings;
+
+/// <summary>
+/// Represents settings related to the user context for a request.
+/// </summary>
+public class RequestUserContextSettings
+{ 
+    public Guid SystemUserId { get; set; }
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Settings/RequestUserContextSettings.cs:8]: Add docs

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Services/IIdentitySecurityTokenService.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    /// <summary>
+    /// Asynchronously creates a new access token.
+    /// </summary>
+    /// <param name="accessToken">The access token to create.</param>
+    /// <param name="saveChanges">Indicates whether changes should be saved to the underlying data store (default is true).</param>
+    /// <param name="cancellationToken">Cancellation token to stop the operation if needed (default is none).</param>
+    /// <returns>A ValueTask representing the asynchronous operation, returning the created AccessToken.</returns>
+    ValueTask<AccessToken> CreateAccessTokenAsync(
+        AccessToken accessToken, 
+        bool saveChanges = true,
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Services/IIdentitySecurityTokenService.cs:19]: Use `CommandOptions`

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Services/IIdentitySecurityTokenService.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    /// <summary>
+    /// Asynchronously creats refresh token for given user
+    /// </summary>
+    /// <param name="refreshToken">The refresh token to create</param>
+    /// <param name="saveChanges">Indicates whether changes should be saved to the underlying data store (default is true)</param>
+    /// <param name="cancellationToken">Cancellation token to stop the operation if needed (default is none)</param>
+    /// <returns>A ValueTask representing the asynchronous operation, returning the created RefreshToken</returns>
+    ValueTask<RefreshToken> CreateRefreshTokenAsync(
+        RefreshToken refreshToken,
+        bool saveChanges = true,
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Services/IIdentitySecurityTokenService.cs:41]: Use `CommandOptions`

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Clients/Mappers/ClientMapper.cs [hunk 6]
```diff
@@ -9,5 +10,7 @@ public class ClientMapper : Profile
     public ClientMapper()
     {
         CreateMap<Client, ClientDto>().ReverseMap();
+        CreateMap<Client, SignUpDetails>().ReverseMap();
+        CreateMap<Client, SignInDetails>().ReverseMap();
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Clients/Mappers/ClientMapper.cs:14]: Remove reverse map option

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/appsettings.Development.json [hunk 7]
```diff
@@ -6,6 +6,26 @@
...
+    "ValidateLifeTime": true,
+    "ValidateIssuerSigningKey": true,
+    "ExpirationTimeInMinutes": 1440,
+    "RefreshTokenExpirationTimeInMinutes": 4320,
+    "RefreshTokenExtendedExpirationTimeInMinutes": 43200
+  },
+  "CacheSettings": {
+    "AbsoluteExpirationInSeconds": 1200,
+    "SlidingExpirationInSeconds": 300
+  },
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/appsettings.Development.json:27]: Move to appsettings instead

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/Configurations/HostConfiguration.Extensions.cs [hunk 8]
```diff
@@ -42,6 +59,70 @@ private static WebApplicationBuilder AddValidators(this WebApplicationBuilder bu
...
+                        ValidateAudience = jwtSettings.ValidateAudience,
+                        ValidateLifetime = jwtSettings.ValidateLifetime,
+                        ValidateIssuerSigningKey = jwtSettings.ValidateIssuerSigningKey,
+                        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwtSettings.SecretKey))
+                    };
+                }
+            );
+        
+        // Register middlewares
+        builder.Services.AddSingleton<AccessTokenValidationMiddleware>();
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/Configurations/HostConfiguration.Extensions.cs:108]: Move to `AddIdentityInfrastructure`

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Commands/SignInCommand.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+/// <summary>
+/// Represents sign in command
+/// </summary>
+public class SignInCommand : ICommand<(AccessToken accessToken, RefreshToken refreshToken)>
+{
+    /// <summary>
+    /// Sign in by email method
+    /// </summary>
+    public SignInDetails? SignInDetails { get; set; }
```
Comment (rashidovasadbek) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Identity/Commands/SignInCommand.cs:15]: must be changed to non-nullable. Remove ? , add defount!

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Persistence/Repositories/AccessTokenRepository.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    /// <param name="cancellationToken">Cancellation token to stop the operation if needed (default is none).</param>
+    /// <returns>A ValueTask representing the asynchronous operation, returning the updated AccessToken.</returns>
+    public async ValueTask<AccessToken> UpdateAsync(AccessToken accessToken, CancellationToken cancellationToken = default)
+    {
+        // Update cache entry with expiration based on AccessToken's ExpiryTime.
+        var cacheEntryOptions = new CacheEntryOptions(accessToken.ExpiryTime - DateTimeOffset.UtcNow, null);
+        await cacheBroker.SetAsync(accessToken.Id.ToString(), accessToken, cacheEntryOptions, cancellationToken);
+
+        return accessToken;
+    }
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Persistence/Repositories/AccessTokenRepository.cs:57]: Remove docs from implementation methods

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Persistence/Repositories/RefreshTokenRepository.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+﻿using Feedback.Analyzer.Domain.Common.Commands;
+using Feedback.Analyzer.Domain.Entities;
+using Feedback.Analyzer.Persistence.Caching.Brokers;
+using Feedback.Analyzer.Persistence.Caching.Models;
+using Feedback.Analyzer.Persistence.Repositories.Interfaces;
+
+namespace Feedback.Analyzer.Persistence.Repositories;
+
+public class RefreshTokenRepository(ICacheBroker cacheBroker) : IRefreshTokenRepository
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Persistence/Repositories/RefreshTokenRepository.cs:9]: Add type docs

**Review by sulton-max — APPROVED** [review]

**Review by isxoqovxasanboy — DISMISSED** [review]

**Review by sulton-max — APPROVED** [review]
Good job

**Review by rashidovasadbek — APPROVED** [review]

**Pull request merged**